File: Monte Carlo Markov chain.R
n=readline(prompt = "Number of states: ")
n=as.numeric(n)

x<-matrix(,nrow=n,ncol=n)
y<-matrix(,nrow=n,ncol=1)
#initial of each state 
for(i in 1:n){
  intit=readline(prompt =paste("The probability of being in state ",i," initially is: "))
  intit<-as.numeric(intit)
  y[i,1]<-intit
}
#each transition probability
for(i in 1:n){
  for(j in 1:n){
    if(i==j){
      prob=readline(prompt =paste("The probability of being in state ",i," is:"))
      prob=as.numeric(prob)
    }else{
      prob=readline(prompt =paste("The probability of transitioning from state ", i," to state ",j," is : "))
      prob=as.numeric(prob)
    }
    x[j,i]<-prob
  }
}
#matrix multiplication
intit<-x
m<-as.numeric(readline(prompt = "Number of transitions: "))
for(num in 1:m-1){
  x<-x%*%intit
}
ans<-x%*%y
for(i in 1:n) print(paste("The probability of being in state ",i," is ",ans[i]))

#Monte Carlo simulation
SIZE=1000000
estimate<-rep(0,5)

for(i in 1:SIZE){
  initial<-sample(1:n,size=1,prob=y[,1])
  for(j in 1:m){
    initial=sample(1:n,size=1,prob=intit[,initial])
  }
  estimate[initial]=estimate[initial]+1
}
options(digit=12)
for(i in 1:n) print(paste("The estimate probability of being in state ",i," is ",estimate[i]/SIZE))

File: Monte Carlo Markov chain.py
import numpy as np
import random
import matplotlib.pyplot as plt

# read user input
n = int(input('Number of states: '))
V = [0] * n
for i in range(n):
    V[i] = float(input(f'The probability of being in state {i + 1} initially is: '))
a = [[0] * n for i in range(n)]
for i in range(n):
    for j in range(n):
        if i == j: a[i][j] = float(input(f'The probability of staying in state {i + 1} is: '))
        else: a[i][j] = float(input(f'The probability of transitioning from state {i + 1} to state {j + 1} is: '))

m = int(input('Number of transitions: '))

# matrix multiplication
t = tm = np.array(a, dtype = float).T
for i in range(m - 1):
    tm = np.dot(tm, t)

res = tm.dot(np.array(V, dtype = float))

print('Results of matrix multiplication')
for i in range(n):
    print('The probability of being in state {} is {:.10f}'.format(i + 1, res[i]))

# monte carlo simulation
SIZE = 100000
v = V[:]
for i in range(m):
    state = 0
    freq = [0] * n
    for j in range(SIZE):
        cur = 0
        s = random.random()
        for k in range(n):
            cur += v[k]
            if s < cur:
                state = k
                break
        
        cur = 0
        s = random.random()
        for k in range(n):
            cur += a[state][k]
            if s < cur:
                state = k
                break
        freq[state] += 1
    for j in range(n):
        v[j] = freq[j] / SIZE

print('Result of Monte Carlo simulation')
for i in range(n):
    print('The probability of being in state {} is {:.10f}'.format(i + 1, v[i]))

# graph
SIZE = 100000
x, y = [[0] * (m + 1) for i in range(n)], [[0] * (m + 1) for i in range(n)]
v = V[:]
for i in range(m+1):
    state = 0
    freq = [0] * n
    for j in range(SIZE):
        cur = 0
        s = random.random()
        for k in range(n):
            cur += v[k]
            if s < cur:
                state = k
                break

        cur = 0
        s = random.random()
        for k in range(n):
            cur += a[state][k]
            if s < cur:
                state = k
                break
        freq[state] += 1
    for j in range(n):
        x[j][i] = i
        y[j][i] = v[j]
        v[j] = freq[j] / SIZE

x = np.array(x)
y = np.array(y)
for i in range(n):
  plt.plot(x[i], y[i])
  plt.ylim(0,1)
  plt.title('State ' + str(i + 1))
  plt.xlabel('Number of transitions')
  plt.ylabel("Probability of being in state " + str(i+1))
  plt.show()
